File: extract.js
const fs = require('fs');
const { exec } = require('child_process');

const begin = new Date();
const fileName = `${begin.getDay()}${begin.getMonth()}${begin.getFullYear()}${begin.getHours()}${begin.getMinutes()}${begin.getSeconds()}`;
console.log(`---------------------------------\nExecution started @${begin.getHours()}:${begin.getMinutes()}:${begin.getSeconds()}`);
exec(`papertrail \'received response for https://api.tap.company/v2/charges\' --min-time \'2022-02-19 08:00:00\' --max-time \'2022-02-21 08:00:00\' > ${fileName}.txt`, (error,output,stderr) => {
    console.log('Command executed.........');
    const end = new Date();
    console.log(`Execution completed @${end.getHours()}:${end.getMinutes()}:${end.getSeconds()}\n---------------------------------`);
    if(error||stderr){
        console.log('Something went wrong');
    }else{
        console.log('Reading file.......\n---------------------------------')
        fs.readFile(`${fileName}.txt`, 'utf8',(error, data) => {
            if(error){
                console.log('Failed to read file');
            }else{
                const extracted = data.match(/([0-9]+\.[0-9]+s)/g);
        
                let timeInFloat = extracted.map((timeInString) => {return parseFloat(timeInString)});
                console.log(`Total : ${timeInFloat.length} records found`);
                console.log(`Minimun : ${Math.min(...timeInFloat)}`);
                const maxValue = Math.max(...timeInFloat);
                console.log(`Maximun : ${maxValue}`);

                var sum = 0;
                timeInFloat.forEach((time) => {sum +=time;});

                console.log(`Average : ${sum/timeInFloat.length}`);

                //Sorting times to get value(s) at/from specific pecentile position
                timeInFloat.sort();
                console.log(`95th Percentile : ${timeInFloat[Math.round((95*timeInFloat.length)/100)]}`);
                console.log(`99th Percentile : ${timeInFloat[Math.round((99*timeInFloat.length)/100)]}`);
                console.log('---------------------------------');
            }
        });        
    }
});

File: extract.go
package main

import (
	"fmt"
	"os/exec"
	"regexp"
	"sort"
	"strconv"
)

func main() {

	//Command with file write is not working. Wirkaroung - use os library and write it to file using write function
	// cmd := exec.Command("papertrail", "error tap gateway timed out", "--min-time", "2022-02-10 00:00:00", "--max-time", "2022-02-16 00:00:00", ">", "testPapertrail.txt")
	cmd := exec.Command("papertrail", "'received response for https://api.tap.company/v2/charges'", "--min-time", "2022-02-19 00:00:00", "--max-time", "2022-02-20 00:00:00")
	fmt.Println(cmd.Args)

	//If you want to store retrieved data in file then un-comment following code to create file
	// file, er := os.Create("sample.txt")
	// if er != nil {
	// 	panic(er)
	// }
	// defer file.Close()

	out, err := cmd.Output()
	if err != nil {
		fmt.Println("Something went wrong : ", err)
	}

	//Un-comment following code to write retrieved data into the created file
	// _, errr := file.Write(out)
	// if err != nil {
	// 	fmt.Println("Something went wrong : ", errr)
	// }

	retrievedData := string(out)

	//Setting a pattern to be searched in retrieved data
	m, _ := regexp.Compile("[0-9]+\\.[0-9]+")

	timesInString := m.FindAllString(retrievedData, -1)

	fmt.Println("Fist : ", timesInString[0])
	fmt.Println("second : ", timesInString[1])
	fmt.Println("third : ", timesInString[2])
	fmt.Println("fourth : ", timesInString[3])

	fmt.Println("Length of timesInString : ", len(timesInString))

	if timesInString != nil {

		fmt.Println("Total : ", len(timesInString))
		// sort.(result)
		var timesInFloat = []float64{}
		sum := 0.0
		for _, timeInString := range timesInString {
			timeInFloat, err := strconv.ParseFloat(timeInString, 64)
			if err != nil {
				fmt.Println("Something went wrong! Failed to covert string into int")
			}
			sum += timeInFloat
			timesInFloat = append(timesInFloat, timeInFloat)
		}

		sort.Float64s(timesInFloat)
		fmt.Println("Minimum : ", timesInFloat[0])
		fmt.Println("Maximum : ", timesInFloat[len(timesInFloat)-1])
		fmt.Println("Avg : ", sum/float64(len(timesInFloat)))
		temp := float64((95 * len(timesInFloat)) / 100)
		fmt.Println("95th Percentile : ", timesInFloat[int(temp)])
		// fmt.Println(result)
	}
}
